src/execution/sql/thread_state_container.cpp

```diff
@@ -83,35 +87,52 @@ void ThreadStateContainer::Reset(const std::size_t state_size, const ThreadState
 }
 
 byte *ThreadStateContainer::AccessCurrentThreadState() {
-  auto &tls_handle = impl_->states_.local();
-  return tls_handle.State();
+  common::SpinLatch::ScopedSpinLatch guard(&impl_->states_latch_);
+  auto iter = impl_->states_.find(std::this_thread::get_id());
+  byte *state = nullptr;
+  if (iter == impl_->states_.end()) {
+    auto *tls_handle = new TLSHandle(this);
+    impl_->states_.insert(std::make_pair(std::this_thread::get_id(), tls_handle));
+    state = tls_handle->State();
+  } else {
+    state = iter->second->State();
+  }
+  return state;
 }
 
 void ThreadStateContainer::CollectThreadLocalStates(std::vector<byte *> *container) const {
   container->clear();
+  common::SpinLatch::ScopedSpinLatch guard(&impl_->states_latch_);
   container->reserve(impl_->states_.size());
   for (auto &tls_handle : impl_->states_) {
-    container->push_back(tls_handle.State());
+    container->push_back(tls_handle.second->State());
   }
 }
 
 void ThreadStateContainer::CollectThreadLocalStateElements(std::vector<byte *> *container,
                                                            const std::size_t element_offset) const {
   container->clear();
+  common::SpinLatch::ScopedSpinLatch guard(&impl_->states_latch_);
   container->reserve(impl_->states_.size());
   for (auto &tls_handle : impl_->states_) {
-    container->push_back(tls_handle.State() + element_offset);
+    container->push_back(tls_handle.second->State() + element_offset);
   }
 }
 
 void ThreadStateContainer::IterateStates(void *const ctx, ThreadStateContainer::IterateFn iterate_fn) const {
+  common::SpinLatch::ScopedSpinLatch guard(&impl_->states_latch_);
   for (auto &tls_handle : impl_->states_) {
-    iterate_fn(ctx, tls_handle.State());
+    iterate_fn(ctx, tls_handle.second->State());
   }
 }
 
 void ThreadStateContainer::IterateStatesParallel(void *const ctx, ThreadStateContainer::IterateFn iterate_fn) const {
-  tbb::parallel_for_each(impl_->states_, [&](auto &tls_handle) { iterate_fn(ctx, tls_handle.State()); });
+  common::SpinLatch::ScopedSpinLatch guard(&impl_->states_latch_);
+  tbb::parallel_for_each(impl_->states_.begin(), impl_->states_.end(), [&](auto &tls_handle) {
```

Comment:
1. Warning: this can lead to dead-lock if the iteration function, for any reason, calls back into `ThreadStateContainer`. In practice, query code only invokes this function after modifications to the states have completed. So, grabbing a lock here isn't strictly necessary.
2. You can remove the `if` clause below since NULL elements should never exist in the map.